crypto/hpke/hpke.c

```diff
@@ -0,0 +1,3137 @@
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Initialise key and IV */
+    if (EVP_DecryptInit_ex(ctx, NULL, NULL, key, iv) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Provide AAD. Can be called zero or more times as required */
+    if (aadlen != 0 && aad != NULL) {
+        if (EVP_DecryptUpdate(ctx, NULL, &len, aad, aadlen) != 1) {
+            OSSL_HPKE_err;
+            goto err;
+        }
+    }
+    /*
+     * Provide the message to be decrypted, and obtain cleartext output.
+     * EVP_DecryptUpdate can be called multiple times if necessary
+     */
+    if (EVP_DecryptUpdate(ctx, plaintext, &len, cipher,
+                          cipherlen - taglen) != 1) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    plaintextlen = len;
+    /* Set expected tag value. Works in OpenSSL 1.0.1d and later */
+    if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG,
+                             taglen, cipher + cipherlen - taglen)) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    /* Finalise decryption.  */
+    if (EVP_DecryptFinal_ex(ctx, plaintext + len, &len) <= 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    if (plaintextlen > *plainlen) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    *plainlen = plaintextlen;
+    memcpy(plain, plaintext, plaintextlen);
+
+err:
+    EVP_CIPHER_CTX_free(ctx);
+    EVP_CIPHER_free(enc);
+    OPENSSL_free(plaintext);
+    return erv;
+}
+
+/*
+ * @brief do AEAD encryption as per the RFC
+ *
+ * @param libctx is the context to use (normally NULL)
+ * @param suite is the ciphersuite
+ * @param key is the secret
+ * @param keylen is the length of the secret
+ * @param iv is the initialisation vector
+ * @param ivlen is the length of the iv
+ * @param aad is the additional authenticated data
+ * @param aadlen is the length of the aad
+ * @param plain is an output
+ * @param plainlen is the length of plain
+ * @param cipher is an output
+ * @param cipherlen input/output, better be big enough on input, exact on output
+ * @return 1 for good otherwise bad
+ */
+static int hpke_aead_enc(OSSL_LIB_CTX *libctx,
+                         OSSL_HPKE_SUITE suite,
+                         unsigned char *key, size_t keylen,
+                         unsigned char *iv, size_t ivlen,
+                         unsigned char *aad, size_t aadlen,
+                         unsigned char *plain, size_t plainlen,
+                         unsigned char *cipher, size_t *cipherlen)
+{
+    int erv = 1;
+    EVP_CIPHER_CTX *ctx = NULL;
+    int len;
+    size_t ciphertextlen;
+    unsigned char *ciphertext = NULL;
+    size_t taglen = 0;
+    uint16_t aead_ind = 0;
+    EVP_CIPHER *enc = NULL;
+    unsigned char tag[16];
+
+    aead_ind = aead_iana2index(suite.aead_id);
+    if (aead_ind == 0) {
+        OSSL_HPKE_err;
+        goto err;
+    }
+    taglen = hpke_aead_tab[aead_ind].taglen;
+    if (taglen != 16) {
```

Comment:
Why this? Why not just confirm `taglen <= sizeof(tag)`?

Reply:
Get rid of the check... This is a fixed table..